### ISSUE TITLE
ERROR running force:source:retrieve: Cannot read property 'getWorkspaceElements' of undefined

### ISSUE THREAD

**Issue by GuntherRoskams** [forcedotcom/cli#383]
### Summary

Error while getting/ updating the metadata, more specific Custom fields

### Steps To Reproduce:

Illuminated cloud DX Project / Retrieve metadata of the projectbased on package.xml

1.  create the XML like below and retrieve the data
```
<?xml version="1.0" encoding="UTF-8"?>
<Package xmlns="http://soap.sforce.com/2006/04/metadata">
    <fullName>SCCConnector</fullName>
    <description>Service Cloud Connector</description>
    <types>
        <members>*</members>
        <name>BusinessProcess</name>
    </types>
    <types>
        <members>*</members>
        <name>CustomApplication</name>
    </types>
    <types>
        <members>*</members>
        <name>ApexClass</name>
    </types>
    <types>
        <members>*</members>
        <name>ApexPage</name>
    </types>
    <types>
        <members>*</members>
        <name>ApexTrigger</name>
    </types>
    <types>
        <members>*</members>
        <name>AuraDefinitionBundle</name>
    </types>
    <types>
        <members>*</members>
        <name>CustomObject</name>
    </types>
    <types>
        <members>*</members>
        <name>CustomField</name>
    </types>
    <types>
        <members>*</members>
        <name>CustomLabel</name>
    </types>
    <types>
        <members>*</members>
        <name>CompactLayout</name>
    </types>
    <types>
        <members>*</members>
        <name>ConnectedApp</name>
    </types>
    <types>
        <members>*</members>
        <name>FlexiPage</name>
    </types>
    <types>
        <members>*</members>
        <name>Layout</name>
    </types>
    <types>
        <members>*</members>
        <name>ListView</name>
    </types>
    <types>
        <members>*</members>
        <name>NamedCredential</name>
    </types>
    <types>
        <members>*</members>
        <name>Profile</name>
    </types>
    <types>
        <members>*</members>
        <name>PlatformCachePartition</name>
    </types>
    <types>
        <members>*</members>
        <name>RecordType</name>
    </types>
    <types>
        <members>*</members>
        <name>StandardValueSet</name>
    </types>
    <types>
        <members>*</members>
        <name>StandardValueSet</name>
    </types>
    <types>
        <members>*</members>
        <name>StaticResource</name>
    </types>
    <types>
        <members>*</members>
        <name>RemoteSiteSetting</name>
    </types>
    <types>
        <members>*</members>
        <name>CustomMetadata</name>
    </types>
    <types>
        <name>CustomField</name>
    </types>
    <version>47.0</version>
</Package>
```
1.  Retrieve the metadata
1. Error received Retrieval Failed Failed to retrieval metadata.Status: FAILEDMessage: TypeError Cannot read property 'getWorkspaceElements' of undefined
1. If you remove the XML node with the CustomField, the metadata will be read correctly

### Expected result

The custom fields loaded into the project list

### Actual result

Error received Retrieval Failed Failed to retrieval metadata.Status: FAILEDMessage: TypeError Cannot read property 'getWorkspaceElements' of undefined

### Additional information

_Feel free to attach a screenshot_.

**SFDX CLI Version**(to find the version of the CLI engine run sfdx --version): sfdx-cli/7.36.0-b1c4e9c7c8 win32-x64 node-v10.15.3

**SFDX plugin Version**(to find the version of the CLI plugin run sfdx plugins --core)
@oclif/plugin-commands 1.2.3 (core)
@oclif/plugin-help 2.2.1 (core)
@oclif/plugin-not-found 1.2.3 (core)
@oclif/plugin-plugins 1.7.8 (core)
@oclif/plugin-update 1.3.9 (core)
@oclif/plugin-warn-if-update-available 1.7.0 (core)
@oclif/plugin-which 1.0.3 (core)
@salesforce/sfdx-trust 3.0.5 (core)
analytics 1.2.1 (core)
generator 1.1.1 (core)
salesforcedx 47.9.0 (core)
├─ salesforcedx-templates 47.9.0 (core)
└─ salesforce-alm 47.10.0 (core)

sfdx-cli 7.36.0 (core)

**OS and version**: Win10.0.18362.778 (Pro)

**Comment by shetzel** [comment]
This was fixed in a recent patch release.  Please update to the latest CLI for the fix.

**Comment by GuntherRoskams** [comment]
Are you sure ?
Just updated Lastest CLI
sfdx-cli/7.54.4-8ff9ba9cc5 win32-x64 node-v10.15.3

Still this issue
17:26	Retrieval Failed
			Failed to retrieval metadata.
			Status: FAILED
			Message: TypeError
			Cannot read property 'getWorkspaceElements' of undefined

**Comment by shetzel** [comment]
I thought that was fixed already but I was wrong.  Apologies!  There is an issue for it here: https://github.com/forcedotcom/cli/issues/345

Your manifest has 2 entries for CustomField.  Are you saying it reproduces when 1 of them is removed or when both are removed?

**Comment by GuntherRoskams** [comment]
While I was logging this issue, I saw indeed 2 entries for the customField metadata, I removed 1, but still the same error.
In my manifest configured in the project, I 've  only 1 entry for the customField

**Comment by shetzel** [comment]
Ok good to know.  Do you have multiple package directories specified in your `sfdx-project.json` or is there just one entry?

**Comment by shetzel** [comment]
If you'd like to help us troubleshoot this since we're not able to easily reproduce it, could you please modify this file around line 177:
C:\Users\<your_windows_username>\AppData\Local\sfdx\client\7.54.4-8ff9ba9cc5\node_modules\salesforce-alm\dist\lib\source\sourceRetrieve.js

If that's not the location then you'll have to run the `source:retrieve` command appending `--dev-debug` and get the location from the debug output.

Replace:
`let filteredElements = se.getWorkspaceElements();`

With:
`let filteredElements;
            if (se) {
                filteredElements = se.getWorkspaceElements();
            }
            else {
                filteredElements = [];
                this.logger.debug(`processResults: key = ${key} NOT FOUND in _ASE`);
                this.logger.debug(`processResults: _ASE = ${_aggregateSourceElements}`);
            }`

Then run the `source:retrieve` command again appending `--dev-debug` and note the additional output.  Also could you check that the correct expected result happened with that change?

**Comment by GuntherRoskams** [comment]
If I replace the code and run the retrieve again, I receive an error
C:/Users/Gunther Roskams/AppData/Local/sfdx/client/7.54.4-8ff9ba9cc5/node_modules/salesforce-alm/dist/lib/source/sourceRetrieve.js:183 this.logger.debug(processResults: key = ${key} NOT FOUND in _ASE); ^^^^^^^^^^^^^^SyntaxError: missing ) after argument list at NativeCompileCache._moduleCompile (C:/Users/Gunther Roskams/AppData/Local/sfdx/client/7.54.4-8ff9ba9cc5/node_modules/v8-compile-cache/v8-compile-cache.js:240:18) at Module._compile (C:/Users/Gunther Roskams/AppData/Local/sfdx/client/7.54.4-8ff9ba9cc5/node_modules/v8-compile-cache/v8-compile-cache.js:186:36) at Function._load (C:/Users/Gunther Roskams/AppData/Local/sfdx/client/7.54.4-8ff9ba9cc5/node_modules/@salesforce/lazy-require/lib/LazyLoader.js:89:24) at require (C:/Users/Gunther Roskams/AppData/Local/sfdx/client/7.54.4-8ff9ba9cc5/node_modules/v8-compile-cache/v8-compile-cache.js:161:20) at Object.<anonymous> (C:/Users/Gunther Roskams/AppData/Local/sfdx/client/7.54.4-8ff9ba9cc5/node_modules/salesforce-alm/dist/lib/source/sourceRetrieveCommand.js:10:26) at Module._compile (C:/Users/Gunther Roskams/AppData/Local/sfdx/client/7.54.4-8ff9ba9cc5/node_modules/v8-compile-cache/v8-compile-cache.js:192:30)

In my sfdx-project file, I've only 1 definition (the standard path force-app)

I tried to retrieve the source with 2 different clients (2 different laptops) and the same package.xml file and the same project. 2 different outcomes :(. On the old created project, I still receive the error, if I create a new project, on another laptop, with the same package XML, I don't have any problem. You told it already, this seems a difficult one to solve. The only workaround for now is 'remove the XML entry to retrieve the CustomField'

**Comment by shetzel** [comment]
The Github markdown changed the code I pasted above.  Those are supposed to be string literals within the debug calls but backticks are used for code formatting.  Let me try again.

```javascript
let filteredElements;
if (se) {
  filteredElements = se.getWorkspaceElements();
} else {
  filteredElements = [];
  this.logger.debug(`processResults: key = ${key} NOT FOUND in _ASE`); 
  this.logger.debug(`processResults: _ASE = ${_aggregateSourceElements}`);
}
```

Thanks for those details about your sfdx-project.json and the difference in projects.  That is all very helpful to get a repro.  There is definitely a difference between a new project and an existing project that contains metadata files locally in terms of what the system "knows about".  Since it doesn't repro in a new project that means there is an interaction with the local metadata and the incoming metadata.

I would love to squash this bug so if you wouldn't mind trying again with the code replacement and reply back with the new output from `--dev-debug` I'd really appreciate it.

**Comment by GuntherRoskams** [comment]
Hi Steve,
The changes solved the problem, unfortunately I couldn't retrieve the debug logs, mentioned in the script. (log was to big)
But for sure, I created a new project with the same org, no problem with the same packageXML. The old project has now the field entries and xml decriptions (with the changes you proposed)

**Comment by shetzel** [comment]
Good to know.  Thank you.  Let's take the conversation back to the earlier issue: https://github.com/forcedotcom/cli/issues/345
